File: backend/app/models/ingredient_price.py
import uuid

from sqlalchemy import (
    Column,
    String,
    DateTime,
    ForeignKey,
    Numeric,
    func,
    Uuid,
    UniqueConstraint,
)

from app.models.user import Base


class UserIngredientPrice(Base):
    __tablename__ = "user_ingredient_prices"

    id = Column(Uuid, primary_key=True, default=uuid.uuid4)
    user_id = Column(Uuid, ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    name = Column(String(120), nullable=False)
    name_normalized = Column(String(120), nullable=False)
    unit = Column(String(20), nullable=False, default="")
    unit_price = Column(Numeric(8, 2), nullable=False)
    last_used_at = Column(DateTime, server_default=func.now(), nullable=False)
    source = Column(String(20), nullable=False, server_default="user_edit")  # user_edit | user_confirm
    created_at = Column(DateTime, server_default=func.now(), nullable=False)

    __table_args__ = (
        UniqueConstraint("user_id", "name_normalized", "unit", name="uq_price_user_name_unit"),
    )

File: app/lib/models/ingredient_price.dart
class IngredientPrice {
  final String id;
  final String name;
  final String unit;
  final double unitPrice;
  final DateTime lastUsedAt;
  final String source; // user_edit | user_confirm

  const IngredientPrice({
    required this.id,
    required this.name,
    required this.unit,
    required this.unitPrice,
    required this.lastUsedAt,
    required this.source,
  });

  factory IngredientPrice.fromJson(Map<String, dynamic> json) => IngredientPrice(
        id: json['id'] as String,
        name: json['name'] as String,
        unit: json['unit'] as String? ?? '',
        unitPrice: (json['unit_price'] as num).toDouble(),
        lastUsedAt: DateTime.parse(json['last_used_at'] as String),
        source: json['source'] as String? ?? 'user_edit',
      );
}
